core/event.cpp

```diff
@@ -93,5 +96,133 @@ Variable broadcast_weights(const DataArrayConstView &events) {
+      events, capacity,
+      overloaded{[](auto &&sparse_, const scipp::index capacity_) {
+                   if (capacity_ > 2 * scipp::size(sparse_))
+                     return sparse_.reserve(capacity_);
+                 },
+                 transform_flags::expect_no_variance_arg<1>,
+                 [](const units::Unit &, const units::Unit &) {}});
+}
+
+namespace filter_detail {
+template <class T>
+using make_select_args = std::tuple<event_list<T>, span<const T>>;
+template <class T, class Index>
+using copy_if_args = std::tuple<event_list<T>, event_list<Index>>;
+
+constexpr auto copy_if = [](const VariableConstView &var,
+                            const VariableConstView &select) {
+  return transform<std::tuple<
+      copy_if_args<double, int32_t>, copy_if_args<float, int32_t>,
+      copy_if_args<int64_t, int32_t>, copy_if_args<int32_t, int32_t>,
+      copy_if_args<double, int64_t>, copy_if_args<float, int64_t>,
+      copy_if_args<int64_t, int64_t>, copy_if_args<int32_t, int64_t>>>(
+      var, select,
+      overloaded{
+          transform_flags::expect_no_variance_arg<1>,
+          [](const auto &var_, const auto &select_) {
+            using VarT = std::decay_t<decltype(var_)>;
+            using Events = event_list<typename VarT::value_type>;
+            const auto size = scipp::size(select_);
+            if constexpr (detail::is_ValuesAndVariances_v<VarT>) {
+              std::pair<Events, Events> out;
+              out.first.reserve(size);
+              out.second.reserve(size);
+              for (const auto i : select_) {
+                out.first.push_back(var_.values[i]);
+                out.second.push_back(var_.variances[i]);
+              }
+              return out;
+            } else {
+              Events out;
+              out.reserve(size);
+              for (const auto i : select_)
+                out.push_back(var_[i]);
+              return out;
+            }
+          },
+          [](const units::Unit &var_, const units::Unit &) { return var_; }});
+};
+
+template <class T>
+const auto make_select = [](const DataArrayConstView &array, const Dim dim,
+                            const VariableConstView &interval) {
+  return transform<
+      std::tuple<make_select_args<double>, make_select_args<float>,
+                 make_select_args<int64_t>, make_select_args<int32_t>>>(
+      array.coords()[dim], subspan_view(interval, dim),
+      overloaded{transform_flags::expect_no_variance_arg<0>,
+                 transform_flags::expect_no_variance_arg<1>,
+                 [](const auto &coord_, const auto &interval_) {
+                   const auto low = interval_[0];
+                   const auto high = interval_[1];
+                   const auto size = scipp::size(coord_);
+                   event_list<T> select_;
+                   for (scipp::index i = 0; i < size; ++i)
+                     if (coord_[i] >= low && coord_[i] < high)
```

Comment:
So we would lose events sitting exactly on the upper boundary at the end of the sorted input event list say?

Reply:
Events at the upper bound are ignored. Whether they are at the end or whether lists are sorted is irrelevant?

This behavior is consistent with histogramming.

Reply:
Note that this is necessary: Otherwise slicing realigned event data would include events in two slices. Intervals need to be `[...)` not `[...]`.